### PR TITLE
[WIP] Different directories for completed/downloading torrents

### PR DESCRIPTION
**Pull request by KristupasSavickas**
TODO: 
- [ ] Saving complete/download directories to session data
- [ ] RPC support
    - [ ] Sycli support

Please let me know what needs to be changed/added for this to be merged in (I'm pretty new to Rust, so I'm not quite confident with my code).

Also, what should I name the new session format version in `session/src/lib.rs`? I tried finding a pattern between the ones used until now, but it looks like random hex strings. Is that the case?

### CODE REVIEW

**Comment by Luminarys** [comment]
I'm a bit busy at the moment, but I'll give this a review over the next day or two.

In general I'm aware that this is a fairly common feature on other clients, but I do wonder if it should be implemented directly in synapse or via an RPC mechanism, i.e. initiate change directory on download start and completion.

The pattern for session IDs is actually git commit short hashes. The whole system could use some rework though.

**Comment by Luminarys** [comment]
This looks reasonable at a first glance, but it's unclear to me why you want to save the paths to session data and also support runtime changes at RPC (if I understand the TODO correctly).

My general opinion of things specified in config files is they should never be allowed to change at runtime, the user should modify the config and restart if it's needed.

File: example_config.toml [hunk 1]
```diff
@@ -35,6 +35,9 @@ bootstrap_node = "router.bittorrent.com:6881"
 session = "~/.local/share/synapse/"
 # Default download directory
 directory = "./"
+# Default directory for completed torrents
+# Empty to use the download directory
+completed = ""
```
Comment (Luminarys) [inline example_config.toml:40]: Won't this set completed to an empty string and cause some sort of path error down the line?

File: sycli/src/main.rs [hunk 2]
```diff
@@ -336,29 +336,12 @@ fn main() {
 
     let client = match Client::new(url.clone()) {
         Ok(c) => c,
-        Err(_) => {
-            eprintln!("Failed to connect to synapse, ensure your URI and password are correct");
+        Err(e) => {
+            eprintln!("Failed to connect to synapse: {}", e);
```
Comment (Luminarys) [inline sycli/src/main.rs:340]: Can you put back in the message to ensure URI/password are correct.
If auth fails a 401 should be sent back, but it won't be really indicated to the user that they may have an incorrect user/password.

File: sycli/src/client.rs [hunk 3]
```diff
@@ -20,7 +21,10 @@ impl Client {
         let mut c = Client {
             ws: client,
             serial: 0,
-            version: Version { major: 0, minor: 0 },
+            version: Version {
```
Comment (Luminarys) [inline sycli/src/client.rs:24]: Good catch, thanks for changing this.

**Comment by KristupasSavickas** [comment]
>  it's unclear to me why you want to save the paths to session data and also support runtime changes at RPC (if I understand the TODO correctly).

> My general opinion of things specified in config files is they should never be allowed to change at runtime, the user should modify the config and restart if it's needed.

The config options are implemented as a "global default". I made it this way so that each torrent could have some degree of customization. This would not be very useful as is, but it would make a lot of sense if we decided to implement a category system in the future (similar to qBittorrent and LabelPlus plugin for Deluge). That way, we could have a hierarchy of paths, e.g. torrent-specific path settings override category path, which override the global path. Of course, if this is ever implemented a more elegant solution would be required.

>  I do wonder if it should be implemented directly in synapse or via an RPC mechanism, i.e. initiate change directory on download start and completion.

I didn't think about the possibility of offloading these tasks to RPC clients (which would not be very practical in my opinion). The clients would have to hold a lot more logic, which would lead to less feature complete clients or even discourage new clients from being made (and native front end client to make this a viable project for power users is required IMHO).

Also, not all RPC clients are running all the time (while the server is), so notifications about completed torrents could not be handled reliably.

**Comment by Luminarys** [comment]
Ok I see what you mean. When I said RPC mechanism, I'm thinking more along the lines of spawning some sort of thread(or even forking) which explicitly connects to the main process via the RPC socket and talks over that, handling all types of non "critical" behavior.

The reason I want this is because the torrent handling code is a complex enough mess as is (I'm sure you've seen this), and I think extracting  logic like this to a separate portion of the codebase will make it easier to reason about things down the line.

Of course one issue with this is how RPC messages should be handled then, since I'm not sure I like implicit communication over the user_data field or related methods. I think for now you can proceed as you're going, though if you do want to think about this it wouldn't be a bad idea.

**Comment by KristupasSavickas** [comment]
In general I think the use of RPC messages should be reserved for remote clients. 

This could be a quick solution, but I don't see it scaling. If I had to guess, the constant serialization of messages would lead high CPU usage and communicating over sockets would stress other parts of the system. There's a big shortage of torrent clients that are fast, feature rich and can handle a lot of torrents (which I'm sure you've noticed). I do see some sense behind this, mainly the fact that only one front end would have to be maintained.

A better way would be to have a level of abstraction, sitting between the "torrent related stuff" and RPC. We could have multiple users of this layer - RPC for remote clients and native callbacks/methods for other non-critical functionality. This could also tie in nicely if a some sort of plugin system is implemented in the future. Also, this would allow Synapse to be built into clients with RPC being an optional feature, sort of like Deluge.

I'll put some thought into this, maybe I'll be able to come up with something elegant.

**Pull request closed without merging**